finality-aleph/src/justification/requester.rs

```diff
@@ -202,6 +203,72 @@ where
     }
 
     pub fn finalized_number(&self) -> NumberFor<B> {
-        self.client.info().finalized_number
+        self.backend.blockchain().info().finalized_number
+    }
+
+    fn do_request(&mut self, hash: &<B as BlockT>::Hash, num: NumberFor<B>) {
+        debug!(target: "aleph-justification",
+               "We have block {:?} with hash {:?}. Requesting justification.", num, hash);
+        self.justification_request_scheduler.on_request_sent();
+        self.block_requester.request_justification(hash, num);
+    }
+
+    // We request justifications for all the children of last finalized block.
+    // Assuming that we request at the same pace that finalization is progressing, it ensures
+    // that we are up to date with finalization.
+    // We also request the child that it's on the same branch as top_wanted since a fork may happen
+    // somewhere in between them.
+    fn request_children(&mut self, info: &Info<B>) {
+        let finalized_hash = info.finalized_hash;
+        let finalized_number = info.finalized_number;
+
+        let children = self
+            .backend
+            .blockchain()
+            .children(finalized_hash)
+            .unwrap_or_default();
+
+        if !children.is_empty() {
+            self.request_status
+                .save_children(finalized_hash, children.len());
+        }
+
+        for child in &children {
+            self.do_request(child, finalized_number);
+        }
+    }
+
+    // This request is important in the case when we are far behind and want to catch up.
+    fn request_wanted(&mut self, mut top_wanted: NumberFor<B>, info: &Info<B>) {
+        let best_number = info.best_number;
+        if best_number <= top_wanted {
+            // most probably block best_number is not yet finalized
+            top_wanted = best_number - NumberFor::<B>::one();
```

Comment:
This could panic if the chain is empty. We should use `saturating_sub`.

Reply:
Fixed.